lib/spack/spack/config.py

```diff
@@ -206,61 +216,340 @@ def __repr__(self):
+        YAML config file that looks like this::
 
+           config:
+             install_tree: $spack/opt/spack
+             module_roots:
+               lmod:   $spack/share/spack/lmod
 
-def highest_precedence_scope():
-    """Get the scope with highest precedence (prefs will override others)."""
-    return list(config_scopes.values())[-1]
+        ``get_config('config')`` will return::
 
+           { 'install_tree': '$spack/opt/spack',
+             'module_roots: {
+                 'lmod': '$spack/share/spack/lmod'
+             }
+           }
 
-def validate_scope(scope):
-    """Ensure that scope is valid, and return a valid scope if it is None.
+        """
+        _validate_section_name(section)
+        merged_section = syaml.syaml_dict()
 
-       This should be used by routines in ``config.py`` to validate
-       scope name arguments, and to determine a default scope where no
-       scope is specified.
+        if scope is None:
+            scopes = self.scopes.values()
+        else:
+            scopes = [self._validate_scope(scope)]
+
+        for scope in scopes:
+            # read potentially cached data from the scope.
+
+            data = scope.get_section(section)
+
+            # Skip empty configs
+            if not data or not isinstance(data, dict):
+                continue
+
+            if section not in data:
+                continue
+
+            merged_section = _merge_yaml(merged_section, data)
+
+        # no config files -- empty config.
+        if section not in merged_section:
+            return {}
+
+        # take the top key off before returning.
+        return merged_section[section]
+
+    def get(self, path, default=None, scope=None):
+        """Get a config section or a single value from one.
+
+        Accepts a path syntax that allows us to grab nested config map
+        entries.  Getting the 'config' section would look like::
+
+            spack.config.get('config')
 
+        and the ``dirty`` section in the ``config`` scope would be::
+
+            spack.config.get('config:dirty')
+
+        We use ``:`` as the separator, like YAML objects.
+
+        TODO: Currently only handles maps. Think about lists if neded.
     """
-    if scope is None:
-        # default to the scope with highest precedence.
-        return highest_precedence_scope()
+        section, _, rest = path.partition(':')
 
-    elif scope in config_scopes:
-        return config_scopes[scope]
+        config = get_configuration()
+        value = config.get_config(section, scope)
+        if not rest:
+            return value
 
-    else:
-        raise ValueError("Invalid config scope: '%s'.  Must be one of %s"
-                         % (scope, config_scopes.keys()))
+        parts = rest.split(':')
+        while parts:
+            key = parts.pop(0)
+            value = value.get(key, default)
+        return value
+
+    def set(self, path, value, scope=None):
```

Comment:
Don't know how much this would be a problem in reality, but I think the semantic of this class is kind of confusing. If I read the code correctly,  by default when we `get` a value we retrieve a merged view of all the scopes. Instead, when we `set` a value, we set it in a particular scope.

Am I wrong saying that this means that:
```python
# We set the value in the highest precedence scope by default
spack.config.set(somepath, a) 

# We merge all the scopes by default
b = spack.config.get(somepath)

# This might be False, e.g. if 'somepath' is a dict spread over scopes
assert a == b
```
might happen? In case is this the behavior we want?

Reply:
This would never be false, unless you specified a scope in the first `spack.config.set`.  FWIW, this new syntax doesn't actually change the semantics of working with the config system.  It just makes it easier to partially update something without getting the entire dictionary for a level.

Reply:
Say we have two scopes for `foo.yaml`:
```yaml
# first scope
foo:
   bar: 5
# second scope
foo:
    baz: 6
```
Am I wrong on the following snippet:
```python
spack.config.set('foo', {'foobar': 6}) 
b = spack.config.get('foo')
assert b == {'foobar': 6, 'baz': 6}
```
?